File: fd-player/src/examples/ghost_bicyle_js/bicycle/chain.js
return (start, end, segments, displacement) => {
  const dx = (end[0] - start[0]) / segments;
  const dy = (end[1] - start[1]) / segments;
  const length = Math.hypot(end[0] - start[0], end[1] - start[1]) || 0.0001;
  const unitsPerLength = segments / length;
  const dispUnits = displacement * unitsPerLength;
  const segmentFraction = 0.2;

  const wrap = (value) => ((value % segments) + segments) % segments;
  const point = (t) => [start[0] + dx * wrap(t), start[1] + dy * wrap(t)];

  return Array.from({ length: segments }, (_, index) => {
    const a = index + dispUnits;
    const b = a + segmentFraction;
    const aw = wrap(a);
    const bw = wrap(b);
    if (bw < aw) {
      return [];
    }
    return {
      type: 'line',
      from: point(a),
      to: point(b),
      stroke: '#fbbf24',
      width: 0.3
    };
  });
};

File: examples/testlib/art/cartoon/bicycle/chain.fs
(start, end, segments, displacement) =>
{
  dx: (end[0] - start[0]) / segments;
  dy: (end[1] - start[1]) / segments;
  length:
    math.Sqrt(
      (end[0] - start[0]) * (end[0] - start[0])
      + (end[1] - start[1]) * (end[1] - start[1]));

  unitsPerLength: if length == 0 then 0 else segments / length;
  dispUnits: displacement * unitsPerLength;
  segmentFraction: 0.2;

  wrap: (value) =>
  {
    q: math.Floor(value / segments);
    r: value - q * segments;
    eval if r < 0 then r + segments else r;
  };

  point: (value) =>
  [
    start[0] + dx * wrap(value),
    start[1] + dy * wrap(value)
  ];

  raw:
    Range(0, segments) map (k, idx) =>
    {
      a: k + dispUnits;
      b: a + segmentFraction;
      aw: wrap(a);
      bw: wrap(b);
      crossed: bw < aw;
      eval if crossed then null else
      {
        type: "line";
        from: point(a);
        to: point(b);
        stroke: "#fbbf24";
        width: 0.3;
      };
    };

  lines: raw filter (x) => x != null;
  eval lines;
}

